### PR TITLE
feat: API Docs

### PR DESCRIPTION
**Pull request by Golodhros**
Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

## Description
chore: moves all 'var' into 'const' across files and docs
[docs: tweaks documentation sidebar element positions!](https://github.com/britecharts/britecharts/commit/9eb48e77e4f37b08a2aa6973c525b157f0fc072a)
[docs: much better api output](https://github.com/britecharts/britecharts/commit/e21bf1c48d2b5dc1d42bb4a521b8bb84c717309f)

## Motivation and Context
Making our new docs homepage
Part of https://github.com/britecharts/britecharts/issues/395

## How Has This Been Tested?
Manually

## Screenshots (if appropriate):
docs-ss

## Types of changes

-   [ ] Refactor (changes the way we code something without changing its functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Latest master code has been merged into this branch
-   [ ] No commented out code (if required, place // TODO above with explanation)
-   [ ] No linting issues
-   [ ] Build is successful
-   [ ] Code follows the [API Guidelines](http://britecharts.github.io/britecharts/topics-index.html#toc5__anchor)
-   [ ] Updated the documentation
-   [ ] Added tests to cover changes
-   [ ] All new and existing tests passed
Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

### CODE REVIEW

File: packages/docs/scripts/template.hbs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{{#module}}
+{{>body~}}
+{{>member-index~}}
+{{>separator~}}
+{{>members~}}
+{{/module}}
```
Comment (gagoar) [inline packages/docs/scripts/template.hbs:6]:
```suggestion
{{/module}}

```

File: packages/docs/scripts/generate-docs.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    //     path.join(__dirname, '../sidebars.js'),
+    //     `module.exports = ${JSON.stringify(sidebars, null, '  ')}`,
+    //     'utf8'
+    // );
+
+    // run prettier on the output json
+    // execSync(`eslint --fix ${path.join(__dirname, '../docs/sidebars.js')}`);
+
+    // Let the user know what step we're on
+    // console.log('\u001B[32m', '✔️  sidebars.js updated', '\u001B[0m');
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:125]: I don't think this part is necessary anymore.

File: packages/docs/scripts/generate-docs.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        });
+
+        // if there's markdown, do stuff
+        if (markdown && markdown.length > 0) {
+            // get the package ID from the file path
+            const matchingExpression = filePath.match(
+                /\/packages\/([\s\S]*?)\/src\/charts\/([\s\S]*?)\//i
+            );
+            const packageId = matchingExpression[1];
+            const chartId = matchingExpression[2];
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:57]:
```suggestion
           const [, packageId, chartId] = matchingExpression;
```

File: packages/docs/scripts/generate-docs.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    // Get the sidebar object
+    let processingPackageName;
+
+    // grab all js files
+    filePaths.forEach((filePath) => {
+        // Generate markdown from JSDoc comments
+        const markdown = jsdoc2md.renderSync({
+            files: filePath,
+            configure: path.join(__dirname, '../jsdoc.conf.json'),
+            // plugin: path.join(__dirname, './template.hbs'),
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:46]:
```suggestion
```

File: packages/docs/scripts/generate-docs.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            'heading-depth': 1,
+        });
+
+        // if there's markdown, do stuff
+        if (markdown && markdown.length > 0) {
+            // get the package ID from the file path
+            const matchingExpression = filePath.match(
+                /\/packages\/([\s\S]*?)\/src\/charts\/([\s\S]*?)\//i
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:54]: could you have it as a const and link to an example that will match the path? just for future self is always nice to find that around a regex :)

File: packages/docs/scripts/generate-docs.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+/**
+ * Runs through packages folders looking for JSDoc and generates markdown docs
+ */
+function generateDocs() {
+    console.log('-= Generating package docs =-');
+    // Use glob to get all js/ts files
+    const pathPattern = path.join(
+        __dirname,
+        '../../core/src/**/*.[jt]s?(x)'
+        // '../packages/**/src/**/*.[jt]s?(x)'
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:23]:
```suggestion
```

File: packages/docs/scripts/generate-docs.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            }
+
+            // get the sub-folder structure relative to /src/
+            let subPath = path
+                .dirname(filePath)
+                .match(/\/src\/charts\/([\s\S]*?)$/i)
+                ? `${
+                      path
+                          .dirname(filePath)
+                          .match(/\/src\/charts\/([\s\S]*?)$/i)[1]
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:75]: you are repeating the regex twice here. could you const the regex and add an example of what should match that?

File: packages/docs/scripts/generate-docs.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    const filePaths = glob.sync(pathPattern, {
+        ignore: [
+            '**/node_modules/**',
+            '**/*.spec.js',
+            '**/*.d.ts',
+            '**/index.js',
+            '**/*.stories.js',
+            '**/*DataBuilder.js',
+            '**/*Data.js',
+        ],
```
Comment (gagoar) [inline packages/docs/scripts/generate-docs.js:34]: this could be a const outside the function.

**Review by gagoar — APPROVED** [review]
no blockers, only nits and thoughts as I was reading the docs generation code.

File: packages/docs/scripts/template.hbs (hunk 1 above)
Comment (Golodhros) [reply packages/docs/scripts/template.hbs:6]: Actually, this is old stuff I am not using. Removing it.

File: packages/docs/scripts/generate-docs.js (hunk 2 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:125]: True, removing.

File: packages/docs/scripts/generate-docs.js (hunk 3 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:57]: +1

File: packages/docs/scripts/generate-docs.js (hunk 4 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:46]: Done

File: packages/docs/scripts/generate-docs.js (hunk 5 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:54]: Great idea

File: packages/docs/scripts/generate-docs.js (hunk 6 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:23]: Done

File: packages/docs/scripts/generate-docs.js (hunk 7 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:75]: Will do

File: packages/docs/scripts/generate-docs.js (hunk 8 above)
Comment (Golodhros) [reply packages/docs/scripts/generate-docs.js:34]: Will do

**Pull request merged**